CON-4857: fixed autocategory assignment in ProductToShop

keulinho opened the pull request.

sbofirov commented:
@keulinho can you try to implement some function which can be used in both places (ProductToShop & ImportService)?

keulinho commented:
implemented some ideas from svens ticket: https://jira.shopware.com/browse/CON-4815 from https://github.com/shopware/SwagConnect/compare/CON-4815-old
and added some tests

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -73,51 +73,47 @@ public function resolve(array $categories)
      *
      * @param array $node
-     * @param null Category $parent
+     * @param string $parentId
```
I guess it must be int

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -73,51 +73,47 @@ public function resolve(array $categories)
      * @return array
      */
-    public function convertTreeToEntities(array $node, Category $parent = null, $categories = [])
+    public function convertTreeToKeys(array $node, $parentId, $returnOnlyLeafs = true, $categories = [])
```
If every public function should be defined in CategoryResolver abstract class as abstract function or implemented there and available for every different implementation of CategoryResolver, otherwise it must be private, but we need it in ImportService.

keulinho replied:
i just move it to the abstract class because there is just this one implementation

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -227,4 +227,23 @@ private function getCategoryData($name)
+
+    /**
+     * @param array $category
+     * @param string $parentId
```
same here, it must be int.

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -227,4 +227,23 @@ private function getCategoryData($name)
+    /**
+     * @param array $category
+     * @param string $parentId
+     * @return Category
```
it returns int

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -126,13 +122,17 @@ public function convertTreeToEntities(array $node, Category $parent = null, $cat
     /**
      * @param array $category
-     * @param Category $parent
+     * @param string $parentId
```
int

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -126,13 +122,17 @@ public function convertTreeToEntities(array $node, Category $parent = null, $cat
         $categoryModel->fromArray($this->getCategoryData($category['name']));
+
+        $parent = $this->categoryRepository->findOneBy([
+            'id' => $parentId
```
you can cast it to int. Useful when someone uses this function with string values. Doctrine won't cast automatically.
I had problems with varchar column in db and int values. then index doesn't work.

sbofirov replied:
If parent doesn't exist then it will throw an exception "Call getId on null", but I think it's fine

sbofirov commented on Components/ImportService.php:
```diff
@@ -240,27 +256,35 @@ public function importRemoteCategory($localCategoryId, $remoteCategoryKey, $remo
+                        ':categoryID' => $category['parentId']
+                    ]
+                );
+                $detailId = $this->manager->getConnection()->fetchColumn(
```
Do we need to fetch detailId? I think we can use only articleId in query below.

sbofirov commented on Components/ImportService.php:
```diff
@@ -240,27 +256,35 @@ public function importRemoteCategory($localCategoryId, $remoteCategoryKey, $remo
+                        VALUES (?, ?, 1)
+                        ON DUPLICATE KEY UPDATE `connect_mapped_category` = 1
+                    ',
+                    [$articleId,  $detailId]
```
will work only with articleId I think

keulinho replied:
no the tests fail then 
serviceContainerTest::testImportCategory

sbofirov commented on Components/CategoryResolver/AutoCategoryResolver.php:
```diff
@@ -73,51 +73,47 @@ public function resolve(array $categories)
+            }
 
-            $remoteCategories = array_merge($remoteCategories, $this->convertTreeToEntities($node['children'], $mainCategory));
+            $remoteCategories = $this->convertTreeToKeys($node['children'], $mainCategory->getId());
```
AutoCategoryResolver still returns only leaf categories.
Won't be possible to use it and assign categories to categories which are not leaf.

keulinho replied:
convert tree to keys has a bool parameter if only leafs will be returned 
per default it is set to true
the import service calls the convertToKeys method directly with this parameter set to false
the resolve function is just called in productToShop and there it is right to return just leafs

sbofirov commented on Components/ProductToShop.php:
```diff
@@ -433,6 +435,20 @@ public function insertOrUpdate(Product $product)
+        //article has to be flushed
+        $categories = $this->categoryResolver->resolve($product->categories);
+        $hasMappedCategory = count($categories) > 0;
+        $detailAttribute->setConnectMappedCategory($hasMappedCategory);
```
there isn't persist of ```detailAttribute```. Also flush was before this change.
If you see on some line below we call ```$this->manager->clear();```.
So this change won't be stored in db.

sbofirov commented on Tests/Legacy/Shopware/Connect/Component/CategoryResolver/AutoCategoryResolverTest.php:
```diff
@@ -78,33 +78,30 @@ public function testResolve()
-        $this->assertNull($this->categoryRepo->findOneByName('Spanish'));
         $this->assertEquals(
             $defaultCategoryId,
             Shopware()->Db()->fetchOne('SELECT parent FROM s_categories WHERE description = ?', ['Adidas'])
```
Please ask Charly about that. In the past we just imported Adidas category into Deutsch

keulinho replied:
in the past Adidas would get parent null and because of that it wasn't visible
it doesn't make sense to import it in the default category because the category names will be in Spanish and that would be pretty confusing

sbofirov commented on Tests/Legacy/Shopware/Connect/Component/ImportServiceTest.php:
```diff
@@ -190,4 +198,34 @@ public function testFindRemoteArticleIdsByCategoryId()
+        $localCategory = $this->categoryRepository->find(14);
+        /** @var \Shopware\CustomModels\Connect\RemoteCategory $remoteCategory */
+        $remoteCategory = $this->remoteCategoryRepository->findOneBy(['categoryKey' => '/deutsch/bücher']);
+
```
not needed empty line

sbofirov commented on Tests/Legacy/Shopware/Connect/Component/ImportServiceTest.php:
```diff
@@ -190,4 +198,34 @@ public function testFindRemoteArticleIdsByCategoryId()
+            'parent' => $localCategory->getId()
+            ]);
+
+        $this->assertTrue($createdLocalCategory!=null);
```
you can use ```$this->assertNotNull``` instead of this or better ```$this->assertInstanceOf()```

sbofirov commented on Tests/Legacy/Shopware/Connect/ProductToShopTest.php:
```diff
@@ -524,12 +529,27 @@ public function testInsertArticleAndAutomaticallyCreateCategories()
             $childCategoryModel->getParent()->getName()
         );
 
+        $categoryRepository = Shopware()->Models()->getRepository('Shopware\Models\Category\Category');
```
please add use statement and call it with ```Category::class```

sbofirov commented on Tests/Legacy/Shopware/Connect/ProductToShopTest.php:
```diff
@@ -524,12 +529,27 @@ public function testInsertArticleAndAutomaticallyCreateCategories()
+        $categoryRepository = Shopware()->Models()->getRepository('Shopware\Models\Category\Category');
+        /** @var \Shopware\Models\Category\Category $childCategoryModel */
+        $childCategoryModel2 = $categoryRepository->findOneBy(['name' => $childCategory2]);
+        $this->assertInstanceOf('Shopware\Models\Category\Category', $childCategoryModel2);
```
Same here

sbofirov commented on Tests/Legacy/Shopware/Connect/ProductToShopTest.php:
```diff
@@ -524,12 +529,27 @@ public function testInsertArticleAndAutomaticallyCreateCategories()
-        $this->assertEquals($childCategory, $assignCategories[0]->getName());
+        $this->assertEquals(2, count($assignCategories));
+        //this category is assigned to the product when it is created
+        $this->assertEquals('Bücher', $assignCategories[0]->getName());
```
why? Probably, because Bücher was mapped to local category in previous test ?

keulinho replied:
no it's hard coded in $this->getProduct()
i just overwrite the categories instead of merge them

sbofirov commented on Components/ConnectFactory.php:
```diff
@@ -201,7 +201,8 @@ public function createSDK()
                 $manager->getRepository('Shopware\CustomModels\Connect\RemoteCategory'),
-                $manager->getRepository('Shopware\CustomModels\Connect\ProductToRemoteCategory')
+                $manager->getRepository('Shopware\CustomModels\Connect\ProductToRemoteCategory'),
+                $manager->getRepository('Shopware\Models\Category\Category')
```
Please replace Shopware\CustomModels\Connect\ProductToRemoteCategory and Shopware\Models\Category\Category with ProductToRemoteCategory::class and Category::class
